### PR TITLE
Preserve changes to frame locals when going up/down in the stack frame

### PR DESCRIPTION
**Pull request by ccordoba12**
## Description

This complements the excellent improvements on PR #13021 by preserving frame locals when going up/down in the stack. Below I show gifs of the previous and new behaviors using Spyder:

**Before**

Old

**After**

New

## Implementation

This was accomplished by saving a copy of all frame locals before the debugger starts and using them to set `curframe_locals` after every command run in the debugger. That prevents that accesses to `f_locals` (which happens in up/down) reset the locals set by the user.

## Notes

* I draw inspiration for this from the patch present in https://bugs.python.org/issue9633, although it's very different.
* There's [PEP 558](https://www.python.org/dev/peps/pep-0558/) that should solve this problem the right way, but currently is marked for Python 3.11 and [an implementation](https://github.com/python/cpython/pull/3640) for it has been four years in the making.

### CODE REVIEW

**Review by MrMino — CHANGES_REQUESTED** [review]
Unfortunately, this causes a regression - it makes it impossible to change the values of the variables in the local scope. The changes are only visible to the debugger, but not to the debugged code.

On `master` branch:
```
Python 3.8.6 (default, Jun 23 2021, 23:50:42) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.0.0.dev -- An enhanced Interactive Python. Type '?' for help.

In [1]: def print_my_var():
   ...:     my_var = '==== unchanged ===='
   ...:     print(my_var)
   ...: 

In [2]: %debug print_my_var()
NOTE: Enter 'c' at the ipdb>  prompt to continue execution.
> <string>(1)<module>()

ipdb> s
--Call--
> <ipython-input-1-5aa28ff21139>(1)print_my_var()
----> 1 def print_my_var():
      2     my_var = '==== unchanged ===='
      3     print(my_var)
      4 

ipdb> n
> <ipython-input-1-5aa28ff21139>(2)print_my_var()
      1 def print_my_var():
----> 2     my_var = '==== unchanged ===='
      3     print(my_var)
      4 

ipdb> n
> <ipython-input-1-5aa28ff21139>(3)print_my_var()
      1 def print_my_var():
      2     my_var = '==== unchanged ===='
----> 3     print(my_var)
      4 

ipdb> my_var = "==== changed ===="
ipdb> n
==== changed ====
--Return--
None
> <ipython-input-1-5aa28ff21139>(3)print_my_var()
      1 def print_my_var():
      2     my_var = '==== unchanged ===='
----> 3     print(my_var)
      4 
```

With the changes presented in this PR:

```
Python 3.8.6 (default, Jun 23 2021, 23:50:42) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.0.0.dev -- An enhanced Interactive Python. Type '?' for help.

In [1]: def print_my_var():
   ...:     my_var = '==== unchanged ===='
   ...:     print(my_var)
   ...: 

In [2]: %debug print_my_var()
NOTE: Enter 'c' at the ipdb>  prompt to continue execution.
> <string>(1)<module>()

ipdb> s
--Call--
> <ipython-input-1-5aa28ff21139>(1)print_my_var()
----> 1 def print_my_var():
      2     my_var = '==== unchanged ===='
      3     print(my_var)
      4 

ipdb> n
> <ipython-input-1-5aa28ff21139>(2)print_my_var()
      1 def print_my_var():
----> 2     my_var = '==== unchanged ===='
      3     print(my_var)
      4 

ipdb> n
> <ipython-input-1-5aa28ff21139>(3)print_my_var()
      1 def print_my_var():
      2     my_var = '==== unchanged ===='
----> 3     print(my_var)
      4 

ipdb> my_var = "==== changed ===="
ipdb> n
==== unchanged ====
--Return--
None
> <ipython-input-1-5aa28ff21139>(3)print_my_var()
      1 def print_my_var():
      2     my_var = '==== unchanged ===='
----> 3     print(my_var)
      4 
```

**Comment by ccordoba12** [comment]
Hey @MrMino, thanks a lot for checking that! I'll try to address that problem when I have some free time.

File: IPython/core/debugger.py [hunk 1]
```diff
@@ -351,6 +354,17 @@ def hidden_frames(self, stack):
             ip_hide = [h if i > ip_start[0] else True for (i, h) in enumerate(ip_hide)]
         return ip_hide
 
+    def preloop(self):
+        """
+        Save a copy of all frame locals so changes to them are not lost
+        when going up/down in the stack frame.
+        """
+        self.frame_locals = [
+            stack_entry[0].f_locals.copy() for stack_entry in self.stack
```
Comment (madbird1304) [inline IPython/core/debugger.py:363]: @ccordoba12 , is there any need to copy this dict? 
If there isn't, removing `copy()` could fix the problem mentioned by @MrMino 
```suggestion
            stack_entry[0].f_locals for stack_entry in self.stack
```

Comment (ccordoba12) [reply IPython/core/debugger.py:363]: Could you clarify? What's a `ChainMap`?

Comment (madbird1304) [reply IPython/core/debugger.py:363]: image

Comment (ccordoba12) [reply IPython/core/debugger.py:363]: Ok, thanks a lot for the suggestion! I'll take a look at it as soon as Io have some free time.

**Comment by Carreau** [comment]
Thanks for your patience and sending this; I was not able to keep up with everything and to keep the number of PRs small I'm going to close this one; If you believe this was a mistake, feel free to comment or re-send it; 

I can't guarantee I'll get to it as I do not have the bandwidth.

**Pull request closed without merging**